frontend/ui/widget/htmlboxwidget.lua

```diff
@@ -283,4 +277,98 @@ function HtmlBoxWidget:onTapText(arg, ges)
     end
 end
 
+function HtmlBoxWidget:setPageNumber(page_number)
+    if page_number == self.page_number then
+        return
+    end
+
+    self.page_number = page_number
+    self.page_boxes = nil
+    self:clearHighlight()
+end
+
+function HtmlBoxWidget:areTextBoxesEqual(boxes1, text1, boxes2, text2)
+    if text1 ~= text2 then
+        return false
+    end
+
+    if boxes1 and boxes2 then
+        if #boxes1 ~= #boxes2 then
+            return false
+        end
+
+        for i = 1, #boxes1, 1 do
+            if boxes1[i] ~= boxes2[i] then
+                return false
+            end
+        end
+
+        return true
+    else
+        return (boxes1 == nil) == (boxes2 == nil)
+    end
+end
+
+-- Returns with true if the highlight has changed.
+function HtmlBoxWidget:clearHighlight()
+    self.hold_start_pos = nil
+    self.hold_end_pos = nil
+    return self:updateHighlight()
+end
+
+-- Returns with true if the highlight has changed.
+function HtmlBoxWidget:updateHighlight()
+    if self.hold_start_pos and self.hold_end_pos then
+        -- getPageText is slow so we only call it when needed, and keep the result.
+        if self.page_boxes == nil then
+            local page = self.document:openPage(self.page_number)
+            self.page_boxes = page:getPageText()
+            page:close()
+        end
+
+        local text_boxes = koptinterface:getTextFromBoxes(self.page_boxes, self.hold_start_pos, self.hold_end_pos)
+        local changed = not HtmlBoxWidget:areTextBoxesEqual(self.highlight_rects, self.highlight_text, text_boxes.boxes, text_boxes.text)
+        if changed then
+            self.highlight_rects = text_boxes.boxes
+            self.highlight_text = text_boxes.text
+        end
+
+        return changed
+    else
+        local changed = self.highlight_rects ~= nil
+        self.highlight_rects = nil
+        self.highlight_text = nil
+        return changed
+    end
+end
+
+function HtmlBoxWidget:redrawHighlight()
+    self:freeBb()
+    UIManager:setDirty("all", "ui", self.dimen)
```

Comment:
I'm glad you commented on this because the whole setDirty call and dialog member was a bit unclear to me. Now it works the same way as in TextBoxWidget.